File: services/py/src/application/limiter/limiter.py
from collections import defaultdict

from src.application.limiter.bucket import TokenBucket
from src.core.config import Config


class RateLimiter:
    def __init__(self, cfg: Config) -> None:
        rate, capacity = cfg.rate_limits.rate, cfg.rate_limits.capacity

        self.per_upstream = cfg.rate_limits.per_upstream
        self.global_bucket = TokenBucket(rate, capacity)
        self.client_buckets = defaultdict(lambda: TokenBucket(rate, capacity))

    async def allow(self, client_ip: str | None = None) -> bool:
        if self.per_upstream and client_ip:
            bucket = self.client_buckets[client_ip]
            return await bucket.consume()
        else:
            return await self.global_bucket.consume()

File: services/go/src/internal/application/limiter/limiter.go
package limiter

import "sync"

type RateLimiter struct {
	buckets  sync.Map
	rate     float64
	capacity float64
}

func NewLimiter(rate, capacity float64) *RateLimiter {
	return &RateLimiter{rate: rate, capacity: capacity}
}

func (l *RateLimiter) Allow(key string) bool {
	val, ok := l.buckets.Load(key)
	if ok {
		return val.(*TokenBucket).Allow()
	}
	tb := NewTokenBucket(l.rate, l.capacity)
	l.buckets.Store(key, tb)
	return tb.Allow()
}
